[WIP] feat: added telemetry for pinecone DB

kartikay-bagla opened the pull request:
Fixes #32.

This is my first foray into NodeJS and its ecosystem so I'm a little unsure about a couple of things:
- ~How do I run tests?~
- ~How do I run the sample app to check if everything is looking good?~ Sample app is working now :)
- ~I have marked a few places in the code where I'm unsure if I need to change code~
- Any general feedback on style/formatting/prose would be very appreciated!

@quest-bot loot #32

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/traceloop/openllmetry-js?pullRequest=39) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/traceloop/openllmetry-js?pullRequest=39) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/traceloop/openllmetry-js?pullRequest=39) it.</sub>

kartikay-bagla commented on packages/instrumentation-pinecone/package.json:
```diff
@@ -0,0 +1,48 @@
+    "@pinecone-database/pinecone": "^2.0.1"
+  },
+  "homepage": "https://github.com/traceloop/openllmetry-js/tree/main/packages/instrumentation-pinecone",
+  "gitHead": "ef1e70d6037f7b5c061056ef2be16e3f55f02ed5"
```
How is this value determined and updated?

nirga replied:
Don't worry about it, it's auto-generated by `lerna` when we publish to `npm`

kartikay-bagla commented on packages/instrumentation-pinecone/src/instrumentation.ts:
```diff
@@ -0,0 +1,143 @@
+    super("@traceloop/instrumentation-pinecone", "0.0.17", config);
+  }
+
+  public manuallyInstrument(
+    module: typeof pinecone & { openLLMetryPatched?: boolean },
+  ) {
+    if (module.openLLMetryPatched) {
+      return;
+    }
+    // TODO: Need to add code here?
+  }
```
From what I can tell in the other instrumentations, the openAI instrumentation also calls the code in the patch method here but the llamaextract instrumentation just leaves this empty. What should I do for pinecone?

kartikay-bagla commented on packages/instrumentation-pinecone/src/instrumentation.ts:
```diff
@@ -0,0 +1,143 @@
+  protected init(): InstrumentationModuleDefinition<any> {
+    const module = new InstrumentationNodeModuleDefinition<any>(
+      "pinecone",
+      [">=2.0.1"],  // TODO: Figure out version here.
```
I'm not exactly clear on what version is supposed to be here. Is it the pinecone library version?

kartikay-bagla commented on packages/sample-app/src/sample_pinecone.ts:
```diff
@@ -0,0 +1,92 @@
+  appName: "sample_pinecone",
+  apiKey: process.env.TRACELOOP_API_KEY,
+  disableBatch: true,
+  instrumentModules: {
+    pinecone: pc_module
+  }
```
Without this, I'm not getting pinecone traces, but the other samples do not manually include instrument modules. What am I missing here?

nirga commented:
Thanks @kartikay-bagla 
Looking into your code now.
Running the tests is easy using `nx` (https://www.traceloop.com/docs/openllmetry/contributing/developing#basic-guide-for-using-nx) - `nx run-many -t test`

nirga commented:
One comment about the test - I'd test it similarly to how we did it in [Python](https://github.com/traceloop/openllmetry/blob/main/packages/traceloop-sdk/tests/test_pinecone_instrumentation.py) (i.e. - instead of made-up vectors).

kartikay-bagla commented:
> One comment about the test - I'd test it similarly to how we did it in [Python](https://github.com/traceloop/openllmetry/blob/main/packages/traceloop-sdk/tests/test_pinecone_instrumentation.py) (i.e. - instead of made-up vectors).

I thought since we're specifically testing the tracing with pinecone, it would be better to reduce dependency on other things.

nirga commented:
> > One comment about the test - I'd test it similarly to how we did it in [Python](https://github.com/traceloop/openllmetry/blob/main/packages/traceloop-sdk/tests/test_pinecone_instrumentation.py) (i.e. - instead of made-up vectors).
> 
> I thought since we're specifically testing the tracing with pinecone, it would be better to reduce dependency on other things.

Yeah that also makes sense so I'm fine with that. There's always the tension between writing integration tests and unit tests. We chose the integration test path since they catch more bugs easily (specifically in the LLM domain where the library maintainers have a tendency to break APIs all the time), but they are harder to run locally and in CI. 
So I'm also fine with us having more easy to run unit tests :) 

BTW we have an [effort on the Python SDK](https://github.com/traceloop/openllmetry/pull/344) to instrument the vectors themselves so might worth doing it here as well.
עע

kartikay-bagla commented:
> BTW we have an [effort on the Python SDK](https://github.com/traceloop/openllmetry/pull/344) to instrument the vectors themselves so might worth doing it here as well.

I'm sorry but I'm not exactly clear on what "instrumenting the vectors" mean.

kartikay-bagla commented on packages/instrumentation-pinecone/test/instrumentation.test.ts:
```diff
@@ -0,0 +1,141 @@
+    provider.addSpanProcessor(new SimpleSpanProcessor(memoryExporter));
+    instrumentation = new PineconeInstrumentation();
+    instrumentation.setTracerProvider(provider);
+    instrumentation.manuallyInstrument(pc_module);
```
@nirga I'm not sure why but I need to call `manuallyInstrument` because otherwise the wrappers are not working. Could you help me debug this?

To replicate this, you can comment out line 42 here and try running the tests.

nirga replied:
@kartikay-bagla you need to initialize the instrumentation in the SDK, [here](https://github.com/traceloop/openllmetry-js/blob/eb364d985e1aeb1d6067f97a520b277f0f797f34/packages/traceloop-sdk/src/lib/tracing/index.ts#L34)

kartikay-bagla replied:
Already did that, check [here](https://github.com/traceloop/openllmetry-js/blob/42165f1b0261c2eda906e945c56394eb3bf1938d/packages/traceloop-sdk/src/lib/tracing/index.ts#L35) and [here](https://github.com/traceloop/openllmetry-js/blob/42165f1b0261c2eda906e945c56394eb3bf1938d/packages/traceloop-sdk/src/lib/tracing/index.ts#L116).

nirga replied:
And it still doesn't work for you without manually instrument?

kartikay-bagla replied:
yep

kartikay-bagla commented:
@nirga I have added span events as well!.

nirga commented:
Thanks @kartikay-bagla! Let me know when it's ready for review and I'll do a full check up :)

kartikay-bagla commented:
> Thanks @kartikay-bagla! Let me know when it's ready for review and I'll do a full check up :)

Apart from needing to call `manuallyInstrument`, everything else is good to go.

nirga commented:
@kartikay-bagla I suspect that the auto-instrumentation is not working well since it can't find the module automatically. We encountered some similar issues in the langchain instrumentation, see the solution @Kartik1397 came up with there - https://github.com/traceloop/openllmetry-js/blob/109cf0efd1452a7b1f5e6eef6dd84efbb86d1622/packages/instrumentation-langchain/src/instrumentation.ts#L38

kartikay-bagla commented:
> @kartikay-bagla I suspect that the auto-instrumentation is not working well since it can't find the module automatically. We encountered some similar issues in the langchain instrumentation, see the solution @Kartik1397 came up with there -
> 
> https://github.com/traceloop/openllmetry-js/blob/109cf0efd1452a7b1f5e6eef6dd84efbb86d1622/packages/instrumentation-langchain/src/instrumentation.ts#L38

I think that fixes it! Sample app is now working as well!

@nirga you can do a full review now.

nirga commented:
Looks good overall @kartikay-bagla! Can we just align the semantic conventions to the ones [we have in Python](https://github.com/traceloop/openllmetry/blob/e81dada24d4a650d6466c6eb6a0a05444aa4740e/packages/opentelemetry-semantic-conventions-ai/opentelemetry/semconv/ai/__init__.py#L57), and add them to the `ai-semantic-conventions` package?

kartikay-bagla commented:
> Looks good overall @kartikay-bagla! Can we just align the semantic conventions to the ones [we have in Python](https://github.com/traceloop/openllmetry/blob/e81dada24d4a650d6466c6eb6a0a05444aa4740e/packages/opentelemetry-semantic-conventions-ai/opentelemetry/semconv/ai/__init__.py#L57), and add them to the `ai-semantic-conventions` package?

@nirga it looks like string formatting isn't supported in typescript (only string interpolation), there are different solutions proposed below, how should I go ahead?
https://stackoverflow.com/questions/20070158/string-format-not-work-in-typescript

Option 1 is to extend the String interface:
```typescript
interface String {
    format(...replacements: string[]): string;
}

if (!String.prototype.format) {
  String.prototype.format = function() {
    var args = arguments;
    return this.replace(/{(\d+)}/g, function(match, number) { 
      return typeof args[number] != 'undefined'
        ? args[number]
        : match
      ;
    });
  };
}
```

Option 2 would be to create a function that can take in format strings with variables and output the result (i.e. same function as above but not part of the string interface)
```typescript
function formatString(a: string, ...replacements: string[]): string {
    return a.replace(/{(\d+)}/g, function (match, number) {
        return typeof replacements[number] != 'undefined'
            ? replacements[number]
            : match
            ;
    });
};
```

Option 3 could be to use wrapper functions around these strings:
```typescript
function myTemplate(a: string, b: string): string {
    return `This is an ${a} for ${a} purposes: ${b}`;
}

alert(myTemplate('example', 'end'));
```

I like Option 2 the best, and if we decide to go ahead with it, how should I implement that?

kartikay-bagla commented:
> Also, I see the test is failing with 404... Does it also happen to you locally?

The 404 could happen due to a missing API key for pinecone. You need to add it to the environment as `PINECONE_API_KEY`

nirga commented:
The API key was there, the problem was with the configuration we had in Pinecone :/ 
I've fixed that now but I see the build is failing as you're missing dependency on ai semantic conventions in the pinecone instrumentation. Also, when I ran the tests locally I get an error:
```
 1 failing
  1) Test Pinecone instrumentation
       should set attributes in span for DB query:
      AssertionError [ERR_ASSERTION]: Expected values to be strictly equal:
2 !== 8

      + expected - actual
      -2
      +8
      
      at Context.<anonymous> (test/instrumentation.test.ts:127:12)
      at processTicksAndRejections (node:internal/process/task_queues:95:5)
```

(BTW the build and test here fails automatically on your commits due to some Github limitation that doesn't allow outside contributors to use our secrets for running the tests and I'm triggering it manually everytime. So it should pass if the tests work and I trigger them).

kartikay-bagla commented:
Fixed package.json issue.

And I have attributed the test error to just not waiting long enough for pinecone to update. I tried running the tests locally about 100 times, it sometimes failed like yours did. Added a 30 second wait time before querying any freshly inserted datam and that seemed to fix the issue!
